### PR TITLE
Refactor dex completion (again)

### LINKED ISSUE #3932: Species that do not count towards the Pokédex should have a flag in SpeciesInfo instead of using a static array
### Description

Context: https://github.com/rh-hideout/pokeemerald-expansion/pull/3900#issuecomment-1876321511

ideally this would deprecate [sRegionalNotCountedList ](https://github.com/Bassoonian/pokeemerald-expansion/blob/0f078ead3077385b7c0c7599d214136cb49aff83/src/birch_pc.c#L24).

### Discord contact info

pkmnsnfrn

### PR DESCRIPTION
**Pull request by Bassoonian**
Refactors the previous solution to the regional dex problem by ignoring all species with the mythical flag instead. Due to Deoxys and Arceus being required for completion in ORAS and PLA, respectively, I added a `mythicalOverride` flag to `gSpeciesInfo` that, when enabled, does in fact require the species to be caught for the dex to be considered complete. Also fixes `HasAllHoennMons` which still had Jirachi and Deoxys hardcoded. I currently left `HasAllKantoMons` as is because people will be repurposing the Hoenn dex for their own region, while Kanto is in this gray zone and hardcoding Mew is (probably?) not a big deal. We could consider updating it and adding functions for all other regions, too, but that's out of scope for this particular PR and should probably go to upcoming.

## Issue(s) that this PR fixes
Fixes #3932 

## **Discord contact info**
bassoonian

### CODE REVIEW

File: src/pokedex.c [hunk 1]
```diff
@@ -4379,12 +4379,12 @@ u16 GetKantoPokedexCount(u8 caseID)
 
 bool16 HasAllHoennMons(void)
 {
-    u16 i;
+    u16 i, j;
```
Comment (AlexOn1ine) [inline src/pokedex.c:4382]: use u32

File: src/birch_pc.c [hunk 2]
```diff
@@ -57,13 +50,18 @@ static const u8 *const sBirchDexRatingTexts[BIRCH_DEX_STRINGS] =
 // This shows your Hoenn Pokedex rating and not your National Dex.
 const u8 *GetPokedexRatingText(u16 count)
 {
-    u32 i;
```
Comment (AlexOn1ine) [inline src/birch_pc.c:60]: can you change the arg to u32?

**Review by AlexOn1ine — DISMISSED** [review]
Shouldn't `mythicalOverride` be added to mythicals in species info?

**Comment by Bassoonian** [comment]
> Shouldn't `mythicalOverride` be added to mythicals in species info?

Nope, because it's game specific and Emerald doesn't distribute Jirachi or Deoxys in-game by default

**Comment by AsparagusEduardo** [comment]
Maybe `dexForceRequired` would be a better name.

**Comment by Bassoonian** [comment]
> Maybe `dexForceRequired` would be a better name.

Done

**Comment by AsparagusEduardo** [comment]
Oh hang on, I just realized.

This whole thing is about National Dex numbers, meaning that it should be tracking `NATIONAL_DEX_x` instead of `SPECIES_x`.
It shouldn't be part of `gSpeciesInfo` conceptually.

File: src/birch_pc.c [hunk 3]
```diff
@@ -55,15 +48,20 @@ static const u8 *const sBirchDexRatingTexts[BIRCH_DEX_STRINGS] =
...
     u16 maxDex = HOENN_DEX_COUNT - 1;
-    for(i = 0; sRegionalNotCountedList[i] != SPECIES_NONE; i++)
+    // doesNotCountForRegionalPokedex
+    for(i = 0; i < HOENN_DEX_COUNT; i++)
     {
-        if (GetSetPokedexFlag(SpeciesToNationalPokedexNum(sRegionalNotCountedList[i]), FLAG_GET_CAUGHT))
-            count--;
-        maxDex--;
+        j = HoennToNationalOrder(i + 1);
+        if (gSpeciesInfo[NationalPokedexNumToSpecies(j)].isMythical && !gSpeciesInfo[NationalPokedexNumToSpecies(j)].dexForceRequired)
```
Comment (AsparagusEduardo) [inline src/birch_pc.c:59]:
```suggestion
        j = NationalPokedexNumToSpecies(HoennToNationalOrder(i + 1));
        if (gSpeciesInfo[j].isMythical && !gSpeciesInfo[j].dexForceRequired)
```

File: src/pokedex.c [hunk 4]
```diff
@@ -4409,7 +4409,7 @@ bool16 HasAllMons(void)
 
     for (i = 1; i < NATIONAL_DEX_COUNT + 1; i++)
     {
-        if (!(gSpeciesInfo[i].isMythical) && !GetSetPokedexFlag(i, FLAG_GET_CAUGHT))
+        if (!(gSpeciesInfo[i].isMythical && !gSpeciesInfo[i].dexForceRequired) && !GetSetPokedexFlag(i, FLAG_GET_CAUGHT))
             return FALSE;
     }
```
Comment (AsparagusEduardo) [inline src/pokedex.c:4414]: you're using nat dex idexes in gSpeciesInfo.

File: include/pokemon.h [hunk 5]
```diff
@@ -394,7 +394,8 @@ struct SpeciesInfo /*0x8C*/
             u32 isPaldeanForm:1;
             u32 cannotBeTraded:1;
             u32 allPerfectIVs:1;
-            u32 padding4:18;
+            u32 dexForceRequired:1; // This mythical species will be taken into account for Pokédex ratings because it is available through normal gameplay
```
Comment (AsparagusEduardo) [inline include/pokemon.h:397]:
```suggestion
            u32 dexForceRequired:1; // This species will be taken into account for Pokédex ratings even if they have the "isMythical" flag set.
```

File: include/birch_pc.h [hunk 6]
```diff
@@ -1,6 +1,6 @@
 #ifndef GUARD_BIRCH_PC_H
 #define GUARD_BIRCH_PC_H
 
-const u8 *GetPokedexRatingText(u16 count);
+const u8 *GetPokedexRatingText(u32 count);
```
Comment (AsparagusEduardo) [inline include/birch_pc.h:4]: afaik, there's no point of changing this to `u32` if you're not changing the variable where this function is called.
In this case, only `gSpecialVar_0x8004` is used, so changing this data type is out of scope.

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: src/pokedex.c (hunk 4 above)
Comment (Bassoonian) [reply src/pokedex.c:4414]: GF started :D what would be better here? There doesn't seem to be an existing function to convert nat dex indexes to species

File: include/birch_pc.h (hunk 6 above)
Comment (AsparagusEduardo) [reply include/birch_pc.h]: Nvm, I was wrong.

File: src/pokedex.c (hunk 4 above)
Comment (Bassoonian) [reply src/pokedex.c:4414]: @AsparagusEduardo 👀

Comment (AsparagusEduardo) [reply src/pokedex.c:4414]: `NationalPokedexNumToSpecies`

Comment (AsparagusEduardo) [reply src/pokedex.c:4414]: > GF started

It was my fault, I'll take the L here xD
image

**Comment by AsparagusEduardo** [comment]
👀 
image

**Comment by Bassoonian** [comment]
> 👀 image

I have absolutely no idea what you're talking about
